### PR TITLE
Add center Loss  Op Support

### PR DESCRIPTION
**Pull request by LLMHao**
Add center Loss op Support .

image

image

### CODE REVIEW

File: paddle/fluid/operators/center_loss_op.cc [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("CentersDiff"),
+                   "Input(CentersDiff) should not be null");
+    PADDLE_ENFORCE(ctx->HasInput(framework::GradVarName("Loss")),
+                   "Input(Loss) should not be null");
+    PADDLE_ENFORCE(ctx->HasOutput(framework::GradVarName("X")),
+                   "Output(X) should not be null");
+
+    // auto x_dims = ctx->GetInputDim("CentersDiff");
```
Comment (guru4elephant) [inline paddle/fluid/operators/center_loss_op.cc]: please remove this comment

File: paddle/fluid/operators/center_loss_op.cu [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    auto centers_dim = centers->dims();
+    auto *out_loss = ctx.Output<Tensor>("Loss");
+    auto loss_data = out_loss->mutable_data<T>(ctx.GetPlace());
+
+    auto *centers_out = ctx.Output<Tensor>("CentersOut");
+    auto *centers_out_data = centers_out->mutable_data<T>(ctx.GetPlace());
+
+    if (centers_out_data != centers_data) {
+      int size = centers_out->numel() * sizeof(T);
+      cudaMemcpyAsync(centers_out_data, centers_data, size,
```
Comment (guru4elephant) [inline paddle/fluid/operators/center_loss_op.cu]: I think this should be equivalent with cudaMemcpy, since you only use a single stream here.

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -352,6 +353,38 @@ def fc(input,
     return helper.append_activation(pre_activation)
 
 
+def center_loss(input,
+                labels,
+                num_classes,
+                alpha,
+                param_attr,
```
Comment (guru4elephant) [inline python/paddle/fluid/layers/nn.py:362]: what is this param_attr used for ?

File: python/paddle/fluid/tests/unittests/test_center_loss.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        self.attrs['need_update'] = True
+        labels = np.array([[0], [2], [0]]).astype(np.int64)
+
+        feat = np.array([[1.0, 2.0, 3.0, 4.0], [9.0, 10.0, 11.0, 12.0],
+                         [1.0, 2.0, 3.0, 4.0]]).astype(np.float32)
+        centers = np.array([[2.0, 3.0, 4.0, 5.0], [5.0, 6.0, 7.0, 8.0],
+                            [7.0, 8.0, 9.0, 10.0],
+                            [13.0, 14.0, 15.0, 16.0]]).astype(np.float32)
+
+        result = np.array([[1.93333, 2.93333, 3.93333, 4.93333],
```
Comment (guru4elephant) [inline python/paddle/fluid/tests/unittests/test_center_loss.py]: can you compute this result with math formula rather than offline computation?

**Comment by guru4elephant** [comment]
also, please update your API.spec since you add new API. and please record your API and op change on inner wiki

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (LLMHao) [reply python/paddle/fluid/layers/nn.py:362]: param_attr is used for initialize the centers use xavier
center_loss=fluid.layers.center_loss(input=pool_out,
                labels=label,
                num_classes=1000,
                alpha=lr_center,
                param_attr=fluid.initializer.Xavier(uniform=False),
                update_center=True)

File: paddle/fluid/operators/center_loss_op.cc (hunk 1 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.cc]: done

File: paddle/fluid/operators/center_loss_op.cu (hunk 2 above)
Comment (gavin1332) [reply paddle/fluid/operators/center_loss_op.cu]: suggest TensorCopy high level api

File: paddle/fluid/operators/center_loss_op.cc [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                   "Output(CentersDiff) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("Loss"),
+                   "Output(Loss) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(
+        ctx->HasOutput("CentersOut"),
+        "Output(CentersOut) of CenterLoss shared data with Centers.");
+
+    ctx->SetOutputDim("CentersDiff", {x_dims[0], product(x_dims) / x_dims[0]});
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.cc]: could the output dimension be the x_dims? Why recalculate

File: paddle/fluid/operators/center_loss_op.cc [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    ctx->SetOutputDim("CentersOut", ctx->GetInputDim("Centers"));
+    ctx->SetOutputDim("Loss", {x_dims[0], 1});
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.cc:57]: what if the input `X` is not a 2D-tensor but a LoDTensor?

File: paddle/fluid/operators/center_loss_op.cc [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    PADDLE_ENFORCE(ctx->HasInput("CenterUpdateRate"),
+                   "Input(CenterUpdateRate) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasInput("Label"),
+                   "Input(Label) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasInput("Centers"),
+                   "Input(Centers) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("CentersDiff"),
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.cc]: i think the **diff** is related to all of the **samples** but not all of the **centers**, so i prefer to `SamplesDiff` or `SampleCenterDiff`

File: paddle/fluid/operators/center_loss_op.cc [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    AddOutput("CentersDiff", "(Tensor) output tensor of center_loss operator.");
+    AddOutput("Loss", "(Tensor) Output tensor of center_loss operator.");
+
+    AddAttr<int>("cluster_num",
+                 "The output cluster num of the center_loss operator.");
+    AddAttr<bool>("need_update", "whether need to update center info.");
+    AddComment(R"DOC(
+**CenterLoss operator**
+implemention of the algorithm in the papper A Discriminative 
+Feature Learning Approach for Deep Face Recognition
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.cc]: the implementation is just a loss, not the full algorithm introduced by that paper. It is better to describe what the loss is and what the feature is (or rather what we use this loss for).

File: paddle/fluid/operators/center_loss_op.cc [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    auto x_dims = ctx->GetInputDim("X");
+    auto x_grad_name = framework::GradVarName("X");
+
+    if (ctx->HasOutput(x_grad_name)) {
+      ctx->SetOutputDim(x_grad_name, x_dims);
+    }
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.cc:117]: As the type is consistent among all the inputs and outputs, so it not required to specify a `GetExpectedKernelType` method. But it is OK to preserve.

File: python/paddle/fluid/layers/nn.py [hunk 10]
```diff
@@ -352,6 +353,38 @@ def fc(input,
...
+                update_center=True):
```
Comment (gavin1332) [inline python/paddle/fluid/layers/nn.py:363]: detailed document is missing here!!! It is very important!
you could refer to [log-loss](https://www.paddlepaddle.org.cn/documentation/docs/zh/1.4/api_cn/layers_cn.html#log-loss)
you should tell user what this operator is and how to use

File: python/paddle/fluid/layers/nn.py [hunk 11]
```diff
@@ -352,6 +353,38 @@ def fc(input,
...
+    helper = LayerHelper('center_loss', **locals())
+    dtype = helper.input_dtype()
+    centers_shape = [num_classes, input.shape[1]]
+    centers_param = helper.create_parameter(
+        attr=helper.param_attr, shape=centers_shape, dtype=dtype)
```
Comment (gavin1332) [inline python/paddle/fluid/layers/nn.py]: the input param_attr is not the helper.param_attr, i think it is what elephant mentioned

File: python/paddle/fluid/layers/nn.py [hunk 12]
```diff
@@ -352,6 +353,38 @@ def fc(input,
     return helper.append_activation(pre_activation)
 
 
+def center_loss(input,
+                labels,
+                num_classes,
+                alpha,
```
Comment (gavin1332) [inline python/paddle/fluid/layers/nn.py:361]: it is better to support both scalar and variable, and if a scalar is specified, we create a variable within this function

File: python/paddle/fluid/tests/unittests/test_center_loss.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        self.op_type = "center_loss"
+        self.dtype = np.float32
+        self.init_dtype_type()
+        batch_size = 6
+        feet_dim = 10
+        cluster_num = 8
+        self.attrs = {}
+        self.attrs['cluster_num'] = cluster_num
+        self.attrs['lambda'] = 0.1
+        self.attrs['need_update'] = True
```
Comment (gavin1332) [inline python/paddle/fluid/tests/unittests/test_center_loss.py]: missing unittest for `need_update` of False, which is an import feature

File: paddle/fluid/operators/center_loss_op.cu [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    diff_xc[index] = X[index] - centers_data[center_idx * x_width + col];
+  }
+}
+
+template <typename T>
+__global__ void UpdateCenters(int cluster_num, const int samples_num,
+                              const int x_width, const int64_t *label,
+                              const T *diff_xc, T *acc_sum, const T *alpha,
+                              T *centers_out) {
+  for (int64_t cidx = blockIdx.x * blockDim.x + threadIdx.x; cidx < cluster_num;
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.cu]: low efficiency. you'd better calculate the scaled different value in each sample, and then updated the center by cudaAtomicAdd or cudaAtomicSub. you can refer to https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/lookup_table_op.cu

File: paddle/fluid/operators/center_loss_op.cu [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    if (centers != centers_out) {
+      framework::TensorCopy(
+          *static_cast<const framework::Tensor *>(centers), ctx_place,
+          *platform::DeviceContextPool::Instance().Get(ctx_place),
+          static_cast<framework::Tensor *>(centers_out));
+    }
+
+    int64_t numel = X->numel();
+
+    ComputeDistance<
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.cu]: ComputeDifferent. distance is always positive

File: paddle/fluid/operators/center_loss_op.h [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    // update centers data
+    if (need_update == true) {
+      for (int i = 0; i < cluster_num; i++) {
+        acc_index = centers_diffacc_data + i * deep_feat_dim;
+        center_index = centers_out_data + i * deep_feat_dim;
+        T scale = alpha / center_update_count[i];
+        blas.SCAL(deep_feat_dim, scale, acc_index);
+        blas.VADD(deep_feat_dim, acc_index, center_index, center_index);
```
Comment (gavin1332) [inline paddle/fluid/operators/center_loss_op.h]: why does not CUDA kernel follow this calculation strategy?
calculate the sample-wise update value, and then apply center updating

**Review by gavin1332 — CHANGES_REQUESTED** [review]
1. missing documents for API
2. re-implement cuda kernel refer to cpu kernel

File: python/paddle/fluid/tests/unittests/test_center_loss.py (hunk 13 above)
Comment (LLMHao) [reply python/paddle/fluid/tests/unittests/test_center_loss.py]: done

File: paddle/fluid/operators/center_loss_op.cc (hunk 7 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.cc]: changed to SampleCenterDiff

File: paddle/fluid/operators/center_loss_op.cu (hunk 14 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.cu]: done

File: paddle/fluid/operators/center_loss_op.cu (hunk 15 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.cu]: done

File: python/paddle/fluid/layers/nn.py (hunk 12 above)
Comment (LLMHao) [reply python/paddle/fluid/layers/nn.py:361]: done.

File: python/paddle/fluid/layers/nn.py (hunk 10 above)
Comment (LLMHao) [reply python/paddle/fluid/layers/nn.py:363]: done

File: paddle/fluid/operators/center_loss_op.h (hunk 16 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.h]: done.

File: python/paddle/fluid/tests/unittests/test_center_loss.py (hunk 4 above)
Comment (LLMHao) [reply python/paddle/fluid/tests/unittests/test_center_loss.py]: done .Add python implemention of center loss in test.

File: python/paddle/fluid/layers/nn.py (hunk 11 above)
Comment (LLMHao) [reply python/paddle/fluid/layers/nn.py]: yes. used to initialize the centers

File: paddle/fluid/operators/center_loss_op.cc (hunk 5 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.cc]: done

File: paddle/fluid/operators/center_loss_op.cc (hunk 8 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.cc]: done

**Comment by renhaosw** [comment]
> 1. missing documents for API
> 2. re-implement cuda kernel refer to cpu kernel

done

**Review by gavin1332 — DISMISSED** [review]
LGTM

**Comment by LLMHao** [comment]
image

**Review by hutuxian — DISMISSED** [review]
LGTM.
The shape of parameter input should be 2D, which will be fixed in a future PR.

**Review by frankwhzhang — DISMISSED** [review]

**Review by gavin1332 — DISMISSED** [review]

**Review by hutuxian — DISMISSED** [review]

**Review by kuke — DISMISSED** [review]
Approve for API addition

File: paddle/fluid/operators/center_loss_op.h [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    T *x_index;
+    T *center_index;
+    T *center_loss_diff_index;
+    T *acc_index;
+    platform::Transform<DeviceContext> trans;
+
+    for (int i = 0; i < batch_size; ++i) {
+      tLabel = label_data[i];
+      center_update_count[tLabel]++;
+      x_index = const_cast<T *>(x_data) + i * deep_feat_dim;  // xi index
```
Comment (sneaxiy) [inline paddle/fluid/operators/center_loss_op.h]: `const_cast` is unnecessary, because it seems that this op would not change `x_data`. Try to make the type of `x_index` to be `const T *` instead of `T*`.

File: paddle/fluid/operators/center_loss_op.h [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+      center_index = const_cast<T *>(centers_data) +
```
Comment (sneaxiy) [inline paddle/fluid/operators/center_loss_op.h]: Op should not change its input. It seems that we can replace `centers_data` with `centers_out_data` and remove `const_cast` here.

File: paddle/fluid/operators/center_loss_op.h (hunk 17 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.h]: done

File: paddle/fluid/operators/center_loss_op.h (hunk 18 above)
Comment (LLMHao) [reply paddle/fluid/operators/center_loss_op.h]: done

**Review by hutuxian — DISMISSED** [review]
LGTM

**Review by sneaxiy — DISMISSED** [review]
LGTM.

**Review by kuke — DISMISSED** [review]

**Review by kuke — APPROVED** [review]

**Review by hutuxian — APPROVED** [review]
